rgw: Add an example for lua scripting (#42169)
ceph/ceph: examples/lua/README.md

> 

Reviewer: please rename this file to nats_adapter.md
in the future we would have other examples here as well

# Introduction

This directory contains examples on how to use [Lua Scripting](https://docs.ceph.com/en/latest/radosgw/lua-scripting/) together with a [NATS Lua client](https://github.com/dawnangel/lua-nats) to add NATS to the list of bucket notifications endpoints.

## NATS
To test your setup:
* Install [NATS](https://docs.nats.io/nats-server/installation) and start a nats-server.

> * Subscribe to the NATS server using [go-nats-examples repo](https://github.com/nats-io/go-nats-examples/releases/tag/0.0.50), choosing the topic to be 'Bucket_Notification' (as defined in the [script]())

Reviewer: ```suggestion
* Subscribe to the NATS server using a [nats subscriber](https://github.com/nats-io/go-nats-examples/tree/master/patterns/publish-subscribe), choosing the topic to be 'Bucket_Notification' (as defined in the [script]())
```

```bash
nats-sub "Bucket_Notification"
```

[Full documentation for subscribing](https://docs.nats.io/nats-server/clients).

Alternatively, configure the script to point to an existing NATS broker by editing the following part in the script to match the parameters of your existing nats server. 

```
local nats_params = {
        host ='{host}',
        port = {port},
}
```

## Usage

* Upload the [script]():

```bash

> radosgw-admin script put --infile=script.lua --context=postRequest

Reviewer: ```suggestion
radosgw-admin script put --infile=nats_adapter.lua --context=postRequest
```

```
* Add the packages used in the script:

```bash
radosgw-admin script-package add --package=nats --allow-compilation
radosgw-admin script-package add --package=lunajson --allow-compilation
radosgw-admin script-package add --package='lua-cjson 2.1.0-1' --allow-compilation
```
* Restart radosgw.
* create a bucket:
```
s3cmd --host=localhost:8000 --host-bucket="localhost:8000/%(bucket)" mb s3://mybucket
```
* upload a file to the bucket and make sure that the nats server received the notification

```
s3cmd --host=localhost:8000 --host-bucket="localhost:8000/%(bucket)" put hello.txt s3://mybucket
```

> Expected output:

Reviewer: please format (some json formatting) the results so its not in one long line

```

## Requirements

> * Lua 5.3

Reviewer: ```suggestion
* Lua 5.3 (or higher)
```

* Luarocks

> * Lua-devel

Reviewer: why do we need lua-devel ?

> Full Installation instructions for [Lua(5.3) and luarocks](https://github.com/luarocks/luarocks/wiki/Installation-instructions-for-Unix). 

Reviewer: don't think this is needed.
lua/luarocks can be installed from official repos for ubuntu and fedora, and from epel on centos
